internal-client/src/main/java/com/tencent/rss/client/impl/grpc/ShuffleServerGrpcClient.java

```diff
@@ -445,17 +449,46 @@ public RssGetShuffleDataResponse getShuffleData(RssGetShuffleDataRequest request
   }
 
   @Override
-  public String getClientInfo() {
-    return "ShuffleServerGrpcClient for host[" + host + "], port[" + port + "]";
-  }
+  public RssGetShuffleIndexResponse getShuffleIndex(RssGetShuffleIndexRequest request) {
+    GetShuffleIndexRequest rpcRequest = GetShuffleIndexRequest
+        .newBuilder()
+        .setAppId(request.getAppId())
+        .setShuffleId(request.getShuffleId())
+        .setPartitionId(request.getPartitionId())
+        .setPartitionNumPerRange(request.getPartitionNumPerRange())
+        .setPartitionNum(request.getPartitionNum())
+        .setLength(request.getLength())
+        .build();
+    long start = System.currentTimeMillis();
+    GetShuffleIndexResponse rpcResponse = blockingStub.getShuffleIndex(rpcRequest);
+    String requestInfo = "appId[" + request.getAppId() + "], shuffleId["
+        + request.getShuffleId() + "], partitionId[" + request.getPartitionId() + "]";
+    LOG.info("GetShuffleIndex for " + requestInfo + " cost " + (System.currentTimeMillis() - start) + " ms");
+
+    StatusCode statusCode = rpcResponse.getStatus();
 
-  private List<BufferSegment> toBufferSegments(List<ShuffleDataBlockSegment> blockSegments) {
-    List<BufferSegment> ret = Lists.newArrayList();
-    for (ShuffleDataBlockSegment segment : blockSegments) {
-      ret.add(new BufferSegment(segment.getBlockId(), segment.getOffset(),
-          segment.getLength(), segment.getUncompressLength(), segment.getCrc(), segment.getTaskAttemptId()));
+    RssGetShuffleIndexResponse response;
+    switch (statusCode) {
+      case SUCCESS:
+        response = new RssGetShuffleIndexResponse(ResponseStatusCode.SUCCESS);
+        if (!rpcResponse.getIndexData().isEmpty()) {
+          response.setShuffleIndexResult(
+              new ShuffleIndexResult(rpcResponse.getIndexData().toByteArray()));
+        }
+        break;
+      default:
+        String msg = "Can't get shuffle index from " + host + ":" + port
+            + " for " + requestInfo + ", errorMsg:" + rpcResponse.getRetMsg();
+        LOG.error(msg);
+        throw new RuntimeException(msg);
```

Comment:
In Colin's pr, RssException replace RuntimeException.

Reply:
I may need to resolve lots of conflics :)

Reply:
Colin's pr hasn't been merged yet.

Reply:
RssException is anthor pr's stuff and no need to introduce it unless that pr merge first and I resovle the conflict.

Reply:
pr has been merged

Reply:
conflicts resolved